### PR TITLE
Add ClipGradByGlobalNormAccuracyCompatible

### PR DESCRIPTION
**Pull request by omoYang**
### PR Category
Operator Mechanism

### PR Types
Others

### Description

Add ClipGradByGlobalNormAccuracyCompatible to make the global norm computation in consistent with torch

### 是否引起精度变化
否

### CODE REVIEW

File: python/paddle/nn/clip.py [hunk 1]
```diff
@@ -643,6 +643,78 @@ def _allow_pure_bf16_global_norm_clip(*args):
...
+        super().__init__()
+        self.clip_norm = float(clip_norm)
+        self.auto_skip_clip = auto_skip_clip
+
+    def _get_global_norm(self, params_grads):
+        local_norm_list = []
+        for p, g in params_grads:
+            if g is None or getattr(p, 'need_clip', True) is False:
+                continue
+            local_norm_list.append(paddle.linalg.vector_norm(g, 2.0))
```
Comment (risemeup1111) [inline python/paddle/nn/clip.py]: P1 **优先级：P1**

这里直接对原始梯度调用 `vector_norm` 会把 SelectedRows/sparse 梯度路径打断。原有 `ClipGradByGlobalNorm._dygraph_clip` 在计算 norm 前会先对 `g.is_selected_rows()` / `SELECTED_ROWS` 做 `merge_selected_rows` 并转成 dense tensor；否则开启 `FLAGS_use_accuracy_compatible_kernel` 后，`paddle.nn.Embedding(..., sparse=True)` 这类已有用例会在全局 norm 计算阶段遇到不支持的稀疏梯度输入。建议复用原路径的 SelectedRows 归并逻辑后再计算 norm，例如：

```suggestion
            merge_grad = g
            if in_dynamic_mode() and g.is_selected_rows():
                merge_grad = merge_selected_rows(g)
                merge_grad = merge_grad._get_tensor_from_selected_rows()
            elif g.type == core.VarDesc.VarType.SELECTED_ROWS:
                merge_grad = merge_selected_rows(g)
                merge_grad = get_tensor_from_selected_rows(merge_grad)
            local_norm_list.append(paddle.linalg.vector_norm(merge_grad, 2.0))
```

File: python/paddle/nn/clip.py [hunk 2]
```diff
@@ -643,6 +643,78 @@ def _allow_pure_bf16_global_norm_clip(*args):
...
+
+        if not local_norm_list:
+            return None
+        global_norm_var = paddle.linalg.vector_norm(
+            paddle.stack(local_norm_list)
+        )
```
Comment (risemeup1111) [inline python/paddle/nn/clip.py]: P1 **优先级：P1**

这里把每个梯度的 `vector_norm` 直接 `stack` 起来，会让混合 dtype 参数组在兼容路径下失效。`paddle.stack` 要求输入 Tensor dtype 一致，而原实现专门把 fp16/bf16、fp32、fp64 norm 分组并按目标 `sum_dtype` cast 后再求和；模型里同时存在 fp16 和 fp32 梯度时，开启 `FLAGS_use_accuracy_compatible_kernel` 会在这里报 dtype 不一致。建议保留原来的 dtype 分组/提升策略，或在 stack 前统一把各个局部 norm cast 到确定的累加 dtype，例如：

```suggestion
        norm_dtype = 'float64' if any(norm.dtype == paddle.float64 for norm in local_norm_list) else 'float32'
        local_norm_list = [
            norm.astype(norm_dtype) if norm.dtype != norm_dtype else norm
            for norm in local_norm_list
        ]
        global_norm_var = paddle.linalg.vector_norm(
            paddle.stack(local_norm_list)
        )
```

File: python/paddle/nn/clip.py [hunk 3]
```diff
@@ -727,6 +799,14 @@ def __str__(self) -> str:
 
     @imperative_base.no_grad()
     def _dygraph_clip(self, params_grads):
+        if paddle.get_flags("FLAGS_use_accuracy_compatible_kernel")[
+            "FLAGS_use_accuracy_compatible_kernel"
+        ]:
+            accuracy_compatible_clip = ClipGradByGlobalNormAccuracyCompatible(
+                self.clip_norm, self.auto_skip_clip
+            )
+            return accuracy_compatible_clip._dygraph_clip(params_grads)
```
Comment (risemeup1111) [inline python/paddle/nn/clip.py]: P1 **优先级：P1**

这里新建的裁剪器只拷贝了 `clip_norm` 和 `auto_skip_clip`，但原 `ClipGradByGlobalNorm` 可能已经被 auto_parallel/sharding 初始化了 `should_comm_on_shard_dim`、`sharding_group`、`mp_group`、`fsdp_group` 等状态。进入兼容路径后这些通信状态全部丢失，原实现中的 all_reduce/reshard 也不会执行，分片参数只会按本 rank 的局部 norm 裁剪，导致分布式训练结果错误。建议不要构造一个丢失状态的新对象；如果必须拆出实现，也要把这些通信属性传递并在兼容路径中执行等价通信逻辑，例如：

```python
accuracy_compatible_clip = ClipGradByGlobalNormAccuracyCompatible(
    self.clip_norm, self.auto_skip_clip
)
for name in (
    'should_comm_on_shard_dim', 'sharding_group', 'mp_group', 'fsdp_group',
    'has_dist_param', 'has_not_dist_param', 'should_comm_on_shard_dim',
):
    if hasattr(self, name):
        setattr(accuracy_compatible_clip, name, getattr(self, name))
```

同时 `_get_global_norm` 里还需要按原路径完成对应 all_reduce/reshard；仅复制属性还不足以修复数值。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
发现几处需要先修复的回归风险，细节已放在行级评论中。主要问题是新的 accuracy-compatible 路径绕过了原有 `ClipGradByGlobalNorm` 对 sparse/SelectedRows、混合 dtype 以及分布式全局 norm 聚合的处理；这些会影响已有训练场景的正确性或可用性。

修复后建议补充对应覆盖：sparse embedding 梯度、fp16/fp32 混合梯度，以及带 sharding/model-parallel 通信状态的全局 norm 裁剪路径。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: python/paddle/nn/clip.py [hunk 4]
```diff
@@ -643,6 +643,116 @@ def _allow_pure_bf16_global_norm_clip(*args):
...
+        # mixed fp16/bf16/fp32 -> fp32
+        return paddle.float32
+
+    def _get_global_norm(self, params_grads):
+        # Group local norms by dtype
+        norm_by_dtype = {}
+        for p, g in params_grads:
+            if g is None or getattr(p, 'need_clip', True) is False:
+                continue
+            norm = paddle.linalg.vector_norm(g, 2.0)
```
Comment (risemeup1111) [inline python/paddle/nn/clip.py:684]: P1 **优先级：P1**

当前提交修复了混合 dtype 聚合，但这里仍然直接对原始 `g` 调用 `vector_norm`，SelectedRows/sparse 梯度路径还会被打断。原有 `ClipGradByGlobalNorm._dygraph_clip` 在计算 norm 前会先对 `g.is_selected_rows()` / `SELECTED_ROWS` 执行 `merge_selected_rows` 并转成 dense tensor；否则开启兼容 flag 后，`paddle.nn.Embedding(..., sparse=True)` 这类已有场景仍会在全局 norm 计算阶段遇到不支持的稀疏梯度输入。建议在计算 norm 前复用原路径的 SelectedRows 归并逻辑：

```suggestion
            merge_grad = g
            if in_dynamic_mode() and g.is_selected_rows():
                merge_grad = merge_selected_rows(g)
                merge_grad = merge_grad._get_tensor_from_selected_rows()
            elif g.type == core.VarDesc.VarType.SELECTED_ROWS:
                merge_grad = merge_selected_rows(g)
                merge_grad = get_tensor_from_selected_rows(merge_grad)
            norm = paddle.linalg.vector_norm(merge_grad, 2.0)
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
复查了新提交：混合 dtype 聚合的直接 `stack` 失败已被处理，但 SelectedRows/sparse 梯度路径仍未合入当前实现，我已在当前 diff 行上补了新的行级评论。分布式全局 norm 聚合的问题也仍由之前的行级线程覆盖，当前提交还没有看到对应修复。

请继续按行级评论处理后再更新。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: python/paddle/nn/clip.py (hunk 2 above)
Comment (omoYang) [reply python/paddle/nn/clip.py]: ok

File: python/paddle/nn/clip.py [hunk 5]
```diff
@@ -643,6 +643,116 @@ def _allow_pure_bf16_global_norm_clip(*args):
...
+
+        if len(all_norms) == 1:
+            combined = all_norms[0]
+        else:
+            combined = paddle.concat(all_norms)
+
+        global_norm_var = paddle.linalg.vector_norm(combined)
+        return global_norm_var
+
+    def _dygraph_clip(self, params_grads):
```
[the remark this answers is not in the document]
Comment (omoYang) [reply python/paddle/nn/clip.py:712]: ok

**Review by wanghuancoder — DISMISSED** [review]

**Review by risemeup1111 — APPROVED** [review]
已复查最新提交。SelectedRows/sparse 梯度以及 PP、cross-mesh、sharding/MP/FSDP 等需要通信聚合的场景现在会回退到原有 `ClipGradByGlobalNorm` 路径，我之前提出的两个阻塞问题已被覆盖；混合 dtype 聚合问题也保持修复。

当前未发现需要继续阻塞合入的问题。CI 仍在运行，最终结果还需要以完整 CI 为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by wanghuancoder — APPROVED** [review]
